File: src/io/github/talkarcabbage/rstimer/newtimers/Standard.java
package io.github.talkarcabbage.rstimer.newtimers;

import java.util.Map;

public class Standard extends NewTimer {
	
	long latestReset;
	long duration;

	public Standard(String name, int tabID, String audio, long latestReset, long duration) {
		super(name, tabID, audio);
		this.latestReset = latestReset;
		this.duration = duration;
	}
	
	/**
	 * Creates a timer from the given map of timer data. Invalid properties will be ignored with a console warning and missing values will be defaulted.
	 * @param dataMap
	 */
	public Standard(Map<String, String> dataMap) {
		super("MISSING", 0, "none"); //Some arbitrary defaults in case of missing data
		latestReset = 0; //Arbitrary default values
		duration = 1000; //Arbitrary default values
		for (Map.Entry<String, String> entry : dataMap.entrySet()) {
			try {
				switch (entry.getKey()) {
				case "name":
					this.name = entry.getValue();
					break;
				case "audio":
					this.audio = entry.getValue();
					break;
				case "tab":
					this.tabID = Integer.parseInt(entry.getValue());
					break;
				case "latestreset":
					this.latestReset = Long.parseLong(entry.getValue());
					break;
				case "duration":
					this.duration = Long.parseLong(entry.getValue());
					break;
				default:
					logger.warning("Unknown property type found while parsing Standard timer:" + entry.getKey());
					break;
				}
			} catch (NumberFormatException e) {
				logger.severe("Invalid timer number value " + entry.getValue() + " for property " + entry.getKey());
			}
		}
	}
	
	public long getLatestReset() {
		return this.latestReset;
	}

	@Override
	public void resetTimer() {
		this.latestReset = System.currentTimeMillis();	
		logger.fine( () -> ("Reset standard timer with data: " + this.toString()));
	}

	@Override
	public void resetTimerComplete() {
		this.latestReset = 0;
		logger.fine( () -> ("Reset-complete standard timer with data: " + this.toString()));
	}

	@Override
	public String getTimerTypeString() {
		return "Standard";
	}

	@Override
	public Map<String, String> getTimerDataMap() {
		Map<String, String> map = super.getTimerDataMap();
		map.put("latestreset", String.valueOf(latestReset));
		map.put("duration", String.valueOf(duration));
		return map;
	}

	@Override
	public String getTooltipText() {
		return "Standard Timer\n" + formatTimeRemaining(getTimeRemaining());
	}

	@Override
	public int getPercentageComplete() {
		long rawPercentage = (100*(System.currentTimeMillis() - latestReset))/(duration);
		return rawPercentage >= 100 ? 100 : (int) rawPercentage;
	}

	@Override
	public long getTimeRemaining() {
		return (latestReset + duration) - System.currentTimeMillis();
	}

}

File: src/main/kotlin/io/github/talkarcabbage/rstimer/newtimers/Standard.kt
package io.github.talkarcabbage.rstimer.newtimers

import java.awt.SystemTray
import java.awt.TrayIcon

class Standard : NewTimer {

	var latestReset: Long = 0
		internal set
	var duration: Long = 0
		internal set

	override val timerTypeString: String
		get() = "Standard"

	override val timerDataMap: Map<String, String>
		get() {
			val map = super.timerDataMap.toMutableMap()
			map["latestreset"] = latestReset.toString()
			map["duration"] = duration.toString()
			return map
		}

	override val tooltipText: String
		get() = "Standard Timer\n"+NewTimer.formatTimeRemaining(timeRemaining)

	override val percentageComplete: Int
		get() {
			val rawPercentage = 100*(System.currentTimeMillis()-latestReset)/duration
			return if (rawPercentage >= 100) 100 else rawPercentage.toInt()
		}

	override val timeRemaining: Long
		get() = latestReset+duration-System.currentTimeMillis()

	constructor(name: String, tabID: Int, audio: Boolean, latestReset: Long, duration: Long) : super(name, tabID, audio) {
		this.latestReset = latestReset
		this.duration = duration
	}

	/**
	 * Creates a timer from the given map of timer data. Data unique to all timers will be
	 * initialized via the super constructor.
	 * @param dataMap
	 */
	constructor(dataMap: Map<String, String>) : super(dataMap) {
		duration = 1000 //Arbitrary default values
		for ((key, value) in dataMap) {
			try {
				when (key) {
					MAP_LATEST_RESET -> this.latestReset = java.lang.Long.parseLong(value)
					MAP_DURATION -> this.duration = java.lang.Long.parseLong(value)
				}
			} catch (e: NumberFormatException) {
				NewTimer.logger.severe("Invalid timer number value $value for property $key")
			}
		}
		if (duration == 0L) {
			duration = 1000
		} //Sanitze

		logger.fine("Initialized a timer: ${this.toString()}")

	}

	override fun onTimerComplete() {
		if (audio) {
			SystemTray.getSystemTray().trayIcons[0]?.displayMessage("RS Timer", "The timer ${this.name} just finished!", TrayIcon.MessageType.INFO)
		}
		if (autoreset) {
			val durationsSinceReset = (System.currentTimeMillis()-latestReset)/duration
			this.latestReset += this.duration*durationsSinceReset //More precise autoresets
		}
	}

	override fun resetTimer() {
		this.latestReset = System.currentTimeMillis()
		NewTimer.logger.fine { "Reset standard timer with data: $this" }
	}

	override fun resetTimerComplete() {
		this.latestReset = 0
		NewTimer.logger.fine { "Reset-complete standard timer with data: $this" }
	}

}
